### PR TITLE
TChannel.is_listening(), automatic listen on advertise, do not allow listen twice.

### PR DESCRIPTION
**Pull request by breerly**
cc @blampe @junchaowu @abhinav

### CODE REVIEW

File: tchannel/tornado/tchannel.py [hunk 1]
```diff
@@ -221,7 +229,16 @@ def listen(self, port=None):
 
         :returns:
             Returns immediately.
+
+        :raises AlreadyListeningError:
+            If listen was already called.
         """
+
+        if self.is_listening():
+            raise AlreadyListeningError(
```
Comment (junchaowu) [inline tchannel/tornado/tchannel.py]: Just throw TChannelError with error message 'already listening'

Comment (breerly) [inline tchannel/tornado/tchannel.py]: No way, I want the client to see the specific exception.

File: tchannel/sync/client.py [hunk 2]
```diff
@@ -94,6 +94,9 @@ def advertise(self, routers, name=None, timeout=None):
         :returns: first advertise result.
         :raises AdvertiseError: when unable to begin advertising.
         """
+        # begin listening if hasn't already
+        if not self.async_client.is_listening():
+            self.async_client.listen()
```
Comment (abhinav) [inline tchannel/sync/client.py]: Is this necessary? Won't the async client do it automatically?

Comment (breerly) [inline tchannel/sync/client.py]: Thanks for catching... I started it here and then moved it to the async client; guess I forgot to remove it from here.

File: tchannel/tornado/tchannel.py [hunk 3]
```diff
@@ -123,6 +124,9 @@ def __init__(self, name, hostport=None, process_name=None,
             for peer_hostport in known_peers:
                 self.peers.add(peer_hostport)
 
+        # server created from calling listening
```
Comment (breerly) [inline tchannel/tornado/tchannel.py]: type

**Comment by breerly** [comment]
@abhinav @blampe I think travis's cache is causing threadloop 0.3.0 to be installed... guessing the check in setup.py, `threadloop>=0.3,<0.4` is being ran against the cache, and `0.3.0` passes so it doesn't bother to go out into `pip` and get `0.3.3`. 

I can either update `setup.py` to be `threadloop>=0.3.3,<0.3` or remove the cache. Which is better / more realitic?

**Comment by breerly** [comment]
@abhinav @blampe hmmm... it actually didn't work even when I tightened up the requirements in setup.py... only when I removed the cache... I'm guessing if the cache causes this type of issue then we want to remove it.

File: tchannel/tornado/tchannel.py (hunk 1 above)
Comment (junchaowu) [inline tchannel/tornado/tchannel.py]: It is weird to call AlreadyListeningError. It could be ServerError, or ServerInitError.

File: .travis.yml [hunk 4]
```diff
@@ -29,14 +29,3 @@ script:
...
   - coveralls -v
 
-#
-# cache dependencies and directories
-# note this can be used on public repos when using container-based infra
-# @see http://docs.travis-ci.com/user/caching/
-# @see http://stackoverflow.com/questions/22024789/speeding-up-travis-ci-dependencies-installation-for-angularjs-project
-# @see https://github.com/rgalanakis/goless/blob/master/.travis.yml
-#
-cache:
```
Comment (blampe) [inline .travis.yml]: How much slower are our runs without this? If it's like... much slower... then we could always clear caches manually in travis if we see this again.

File: tchannel/tornado/tchannel.py (hunk 1 above)
Comment (blampe) [inline tchannel/tornado/tchannel.py]: Agree with @breerly, `ServerError` is way too generic.

**Comment by blampe** [comment]
Looks like tests are still about a minute so :+1:

**Comment by abhinav** [comment]
LGTM. :shipit:

**Comment by breerly** [comment]
This new Github css looks great

File: .travis.yml (hunk 4 above)
Comment (breerly) [inline .travis.yml]: Looks like not too much slower

**Pull request merged**